src/Core/Category/NameBuilder/CategoryDisplayNameBuilder.php

```diff
@@ -0,0 +1,183 @@
+     *
+     * @param CategoryId $categoryId
+     * @param string $categoryName
+     * @param ShopId $shopId
+     * @param LanguageId $languageId
+     * @param string[] $breadcrumbParts
+     * @param bool $useCache
+     *
+     * @return string
+     */
+    public function buildWithBreadcrumbs(
+        CategoryId $categoryId,
+        string $categoryName,
+        ShopId $shopId,
+        LanguageId $languageId,
+        array $breadcrumbParts,
+        bool $useCache = true
+    ): string {
+        return $this->buildDisplayName($categoryId, $categoryName, $shopId, $languageId, $breadcrumbParts, $useCache);
+    }
+
+    /**
+     * @param ShopId $shopId
+     * @param LanguageId $langId
+     *
+     * @return string
+     */
+    private function buildCacheKey(ShopId $shopId, LanguageId $langId): string
+    {
+        return sprintf(
+            'Category::duplicateCategoryNames_shop_%s_lang_%s',
+            $shopId->getValue(),
+            $langId->getValue()
+        );
+    }
+
+    /**
+     * @param ShopId $shopId
+     * @param LanguageId $languageId
+     * @param bool $useCache
+     *
+     * @return string[]
+     */
+    private function getDuplicateNames(ShopId $shopId, LanguageId $languageId, bool $useCache): array
+    {
+        if (!$useCache) {
+            return $this->categoryRepository->getDuplicateNames($shopId, $languageId);
+        }
+
+        $cacheKey = $this->buildCacheKey($shopId, $languageId);
+
+//      @todo: consider using Symfony\Component\Cache\Adapter\AdapterInterface instead of legacy Cache
+        if (Cache::isStored($this->buildCacheKey($shopId, $languageId))) {
+            return Cache::retrieve($cacheKey);
+        }
+
+        $duplicateNames = $this->categoryRepository->getDuplicateNames($shopId, $languageId);
+        Cache::store($cacheKey, $duplicateNames);
+
+        return $duplicateNames;
+    }
+
+    /**
+     *  If there are multiple categories with identical names, we want to be able to tell them apart,
+     *  so we use breadcrumb path instead of category name.
+     *  However, whole breadcrumb path would probably be too long, therefore not UX friendly.
+     *  Calculating "optimal" breadcrumb length seems too complex compared to the value it could bring.
+     *  So, we show one parent name and category name, as it is simple and should cover most cases.
+     *
+     * e.g. "Clothes > Women"
+     *
+     * @param CategoryId $categoryId
+     * @param string $categoryName
+     * @param ShopId $shopId
+     * @param LanguageId $languageId
+     * @param array|null $breadcrumbParts
+     * @param bool $useCache
+     *
+     * @return string
+     */
+    private function buildDisplayName(
+        CategoryId $categoryId,
+        string $categoryName,
+        ShopId $shopId,
+        LanguageId $languageId,
+        ?array $breadcrumbParts = null,
+        bool $useCache = true
+    ): string {
```

Comment:
```suggestion
    private function buildDisplayName(
        string $categoryName,
        ShopId $shopId,
        LanguageId $languageId,
        CategoryId $categoryId = null,
        ?array $breadcrumbParts = null,
        bool $useCache = true
    ): string {
```

Reply:
`$categoryId` and `$breadcrumbParts` are both optional since they both correspond to two different use cases:
- `build` : no breadcrumb provided, so we need the category ID to get it via the repository
- `buildWithBreadcrumbs`: breadcrumb is provided so no need for the category ID